snakemake_storage_plugin_xrootd/__init__.py

```diff
@@ -66,128 +78,188 @@ class StorageProviderSettings(StorageProviderSettingsBase):
-        default_factory=dict,
+    url_decorator: Optional[Callable] = field(
+        default=lambda x: x,
         metadata={
-            "help": "Optional query parameters for XRootD (e.g. xrd.wantprot=unix, "
-            "authz=XXXXXX)",
-            "env_var": True,
-            # Optionally specify a function that parses the value given by the user.
-            # This is useful to create complex types from the user input.
-            "parse_func": parse_query_params,
-            # If a parse_func is specified, you also have to specify an unparse_func
-            # that converts the parsed value back to a string.
-            "unparse_func": unparse_query_params,
+            "help": "A function to decorate the URL e.g. wrapping an auth token.",
+            "env_var": False,
             "required": False,
-            "nargs": "+",
         },
     )
 
 
-# Required:
-# Implementation of your storage provider
-# This class can be empty as the one below.
-# You can however use it to store global information or maintain e.g. a connection
-# pool.
 class StorageProvider(StorageProviderBase):
-    # For compatibility with future changes, you should not overwrite the __init__
-    # method. Instead, use __post_init__ to set additional attributes and initialize
-    # futher stuff.
-
     def __post_init__(self):
-        # This is optional and can be removed if not needed.
-        # Alternatively, you can e.g. prepare a connection to your storage backend here.
-        # and set additional attributes.
-
-        userpass = ""
-        if self.settings.username:
-            userpass += self.provider.settings.username
-        if self.settings.password:
-            userpass += f":{self.settings.password}"
-        if userpass:
-            userpass += "@"
-        port = f":{self.settings.port}" if self.settings.port else ""
-        self.netloc = f"{userpass}{self.settings.host}{port}"
-
-        self.filesystem_client = client.FileSystem(f"root://{self.netloc}")
-        self.file_client = client.File()
+        self.username = self.settings.username
+        self.password = self.settings.password
+        self.host = self.settings.host
+        self.port = self.settings.port
+        self.url_decorator = self.settings.url_decorator
+        # List of error codes that there is no point in retrying
+        self.no_retry_codes = [
+            3000,
+            3001,
+            3002,
+            3006,
+            3008,
+            3009,
+            3010,
+            3011,
+            3013,
+            3017,
+            3021,
+            3025,
+            3030,
+            3031,
+            3032,
+        ]
+
+    def _check_status(self, status: XRootDStatus, error_preamble: str):
+        if not status.ok:
+            if status.errno in self.no_retry_codes:
+                raise XRootDFatalException(f"{error_preamble}: {status.message}")
+            raise OSError(f"{error_preamble}: {status.message}")
```

Comment:
When raising errors, use the following rule of thumb:
* if they are exposed to the user, raise a WorkflowError("...").
* otherwise, raise whatever error is appropriate, but make sure it is handled in a calling method so that it never is exposed to the user.

Reply:
Okay, thanks! I've changed the `OSError`s to `WorkflowError`s